src/osd/PrimaryLogPG.cc

```diff
@@ -2225,6 +2225,17 @@ void PrimaryLogPG::do_op(OpRequestRef& op)
       return;
     }
 
+    if (auto blocked_iter = objects_blocked_on_unreadable_adjacent_snap.find(head);
+	blocked_iter != std::end(objects_blocked_on_unreadable_adjacent_snap)) {
+      hobject_t to_wait_on(head);
+      to_wait_on.snap = blocked_iter->second;
+      if (is_unreadable_object(to_wait_on)) {
+	wait_for_unreadable_object(to_wait_on, op);
+	return;
+      } 
+      objects_blocked_on_unreadable_adjacent_snap.erase(head);
```

Comment:
Is this actually possible?  Presumably, we'd have cleared the objects_blocked_on_unreadable_adjacent_snap and updated the state for is_unreadable_object atomically.

Reply:
Hm.. objects_blocked_on_unreadable_adjacent_snap---this is newly added by this commit---would not be cleared until the adjacent snapshot is recovered. Am I missing something?

Reply:
Reaching line 2236 means that the check in 2232 is false.  Thus, we already recovered the adjacent snapshot.  How is it possible that head is still in objects_blocked_on_unreadable_adjacent_snap?

Reply:
objects_blocked_on_unreadable_adjacent_snap is not cleared even though the adjacent snapshot is recovered---I didn't add such operation on the recovery procedure.

Therefore, if the adjacent snapshot is recovered, the OP is re-queued, at which point moreover the head is still in objects_blocked_on_unreadable_adjacent_snap. Then, the program counter reaches line 2236 again. 

Do we need to clear objects_blocked_on_unreadable_adjacent_snap when the adjacent snapshot is readable, like finish_degraded_object()?

Reply:
Done.